### PR TITLE
[Example] Adding error_handling & error_details example

### PR DESCRIPTION
**Pull request by veblush**
Adding a new example showing how to propagate error status from the server to the client with optional detail information.

### CODE REVIEW

File: examples/cpp/error_handling/greeter_error_client.cc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+              << std::endl;
+    stub_->async()->SayHello(&context, &request, &reply,
+                             [&mu, &cv, &done, &status](Status s) {
+                               status = std::move(s);
+                               std::lock_guard<std::mutex> lock(mu);
+                               done = true;
+                               cv.notify_one();
+                             });
+    std::unique_lock<std::mutex> lock(mu);
+    while (!done) {
```
Comment (eugeneo) [inline examples/cpp/error_handling/greeter_error_client.cc:78]: Suggestion: `mu.LockWhen` would remove the need for a separate condition variable, streamlining the code.

File: examples/cpp/error_handling/greeter_error_client.cc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ABSL_FLAG(std::string, target, "localhost:50051", "Server address");
+
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::Status;
+using helloworld::Greeter;
+using helloworld::HelloReply;
+using helloworld::HelloRequest;
+
+class GreeterClient {
```
Comment (eugeneo) [inline examples/cpp/error_handling/greeter_error_client.cc:47]: Suggestion: change this to a SayHello function that takes a stub as an argument.

**Review by eugeneo — APPROVED** [review]
The comments are suggestions, mostly trying to make the example shorter so it is easier to follow.

File: examples/cpp/error_handling/greeter_error_client.cc (hunk 1 above)
Comment (eugeneo) [reply examples/cpp/error_handling/greeter_error_client.cc:78]: Ignore this comment, this is `std` and not `ABSL`

File: examples/cpp/error_handling/greeter_error_client.cc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    ClientContext context;
+    // The actual RPC.
+    std::mutex mu;
+    std::condition_variable cv;
+    bool done = false;
+    Status status;
+    std::cout << absl::StrFormat("### Send: SayHello(name=%s)", user)
+              << std::endl;
+    stub_->async()->SayHello(&context, &request, &reply,
+                             [&mu, &cv, &done, &status](Status s) {
```
[the remark this answers is not in the document]
Comment (veblush) [reply examples/cpp/error_handling/greeter_error_client.cc:71]: Done.

Code style guide says that `Use default capture by reference ([&]) only when the lifetime of the lambda is obviously shorter than any potential captures.`. In this case, `[&]` can be used here.

File: examples/cpp/error_handling/greeter_error_server.cc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  ServerUnaryReactor* SayHello(CallbackServerContext* context,
+                               const HelloRequest* request,
+                               HelloReply* reply) override {
+    ServerUnaryReactor* reactor = context->DefaultReactor();
+    // Checks the length of name
+    if (request->name().empty() || request->name().length() > 10) {
+      // Returns an error status with error code and message.
+      reactor->Finish(Status(StatusCode::INVALID_ARGUMENT,
+                             "Length of `Name` should be between 1 and 10"));
+      return reactor;
```
[the remark this answers is not in the document]
Comment (veblush) [reply examples/cpp/error_handling/greeter_error_server.cc:63]: Done.

File: examples/cpp/error_handling/greeter_error_client.cc (hunk 2 above)
Comment (veblush) [reply examples/cpp/error_handling/greeter_error_client.cc]: I see where this is coming from but I've decided to keep this structure to be consistent with other examples.

**Comment by veblush** [comment]
I just realized that I need to split this one example into two examples; one for a simple error handling and the other for a complicate one demonstrating how to use error details.

**Pull request merged**